### PR TITLE
NEW: Implement unshift() in field list classes (closes #4834)

### LINKED ISSUE #4834: FieldList, CompositeField, TabSet etc should implement unshift()
We have `ArrayList::unshift()` that could be useful in the context of forms/fields.

`FieldList` will currently *mostly* work, because it inherits it from `ArrayList::unshift()`, but it should call the same extra methods that `FieldList::push()` does.

### PR DESCRIPTION
**Pull request by kinglozzer**

### CODE REVIEW

File: forms/FieldList.php [hunk 1]
```diff
@@ -412,16 +412,27 @@ public function insertAfter($name, $item) {
...
 
 	/**
-	 * Push a single field into this FieldList instance.
+	 * Push a single field onto the end of this FieldList instance.
 	 *
 	 * @param FormField $item The FormField to add
-	 * @param string $key An option array key (field name)
 	 */
-	public function push($item, $key = null) {
+	public function push($item) {
```
Comment (kinglozzer) [inline forms/FieldList.php]: No BC issues here - this param isn’t used (in this class, or in the parent `ArrayList` class), and won’t error if passed anyway or added in a subclass

Comment (tractorcow) [inline forms/FieldList.php]: Hm, it looks ok to me.... parent::push() doesn't have a second parameter anyway.

File: forms/FieldList.php [hunk 2]
```diff
@@ -412,16 +412,27 @@ public function insertAfter($name, $item) {
...
-		return parent::push($item, $key);
+		return parent::push($item);
+	}
+
+	/**
+	 * Push a single field onto the beginning of this FieldList instance.
+	 *
+	 * @param FormField $item The FormField to add
+	 */
+	public function unshift($item) {
```
Comment (tractorcow) [inline forms/FieldList.php]: Why not use a strict typehint as you did in FormField::unshift?

**Comment by dhensby** [comment]
Can we get this as part of a list interface for master, please?

**Comment by tractorcow** [comment]
We'll have to change the typehint from `FormField` to `mixed` but I agree it's good addition.

**Pull request merged**